trace/trace.go

```diff
@@ -0,0 +1,68 @@
+package trace
+
+import (
+	"time"
+)
+
+////////////////////////////////////////////////////////////////////////////////
+
+type ClusterTrace struct {
+	// TopoChanged is called when the cluster's topology changed.
+	TopoChanged func(ClusterTopoChanged)
+	// GotResponse is called after the radix.Do executed.
+	GotResponse func(ClusterGotResponse)
+}
+
+type ClusterNodeInfo struct {
+	Addr      string
+	Slots     [][2]uint16
+	IsPrimary bool
+}
+
+type ClusterTopoChanged struct {
+	Added   []ClusterNodeInfo
+	Removed []ClusterNodeInfo
+}
+
+type ClusterGotResponse struct {
+	Addr       string
+	Key        string
+	RetryCount int
+	Err        error
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+type PoolTrace struct {
+	// ConnectDone is called when a new connection's Dial completes. The provided
+	// Err indicates whether the connection successfully completed.
+	ConnectDone func(PoolConnectDone)
+	// ConnClosed is called before closing the connection.
+	ConnClosed func(PoolConnClosed)
+}
+
+type PoolInfo struct {
+	Addr string
+}
+
+type PoolConnInfo struct {
+	PoolSize   int
+	BufferSize int
+	AvailCount int
+}
+
+type PoolConnectDone struct {
+	PoolInfo
+	PoolConnInfo
+	Reason      string
```

Comment:
These `Reason` fields are going to need to be enumerated as consts in this package, as their own type:

```
type PoolConnectReason string

const (
    PoolConnectReasonInitialize PoolConnectReason = "initialize"
    PoolConnectReasonRefill PoolConnectReason = "refill"
    ....
)
```

You can leave the "TEST" reasons as-is, cause tests don't matter.

Reply:
That's the idea what I want to know. Managing reasons in a separate package are fine.